src/librustc/middle/const_eval.rs

```diff
@@ -916,7 +916,16 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
         }
       }
       hir::ExprPath(..) => {
-          let opt_def = tcx.def_map.borrow().get(&e.id).map(|d| d.full_def());
+          let opt_def = if let Some(def) = tcx.def_map.borrow().get(&e.id) {
+              // FIXME: There's probably a better way to make sure we don't
+              // panic here.
+              if def.depth != 0 {
+                  signal!(e, NonConstPath);
```

Comment:
You could extend the error enum to have a specific variant for this situation, as `NonConstPath` signifies a valid but not constant path imo